### PR TITLE
3932-ECRPublicFetcher

### PR DESCRIPTION
**Pull request by ofiriro3**

### CODE REVIEW

File: resources/fetchers/ecr_fetcher_test.go [hunk 1]
```diff
@@ -155,28 +212,34 @@ func (s *ECRFetcherTestSuite) TestCreateFetcher() {
...
 		ecrProvider.EXPECT().DescribeRepositories(mock.Anything, mock.MatchedBy(func(repo []string) bool {
-			return reflect.DeepEqual(test.expectedRepositoryNames, repo)
-		})).Return(test.expectedRepository, nil)
-
-		privateRepoRegex := fmt.Sprintf(testPrivateRepositoryTemplate, test.identityAccount, test.region)
-		//Maybe will need to change this texts
-		regexTexts := []string{
-			privateRepoRegex,
-			testPublicRepositoryRegex,
+			return reflect.DeepEqual(test.expectedRepositoriesNames, repo)
```
Comment (amirbenun) [inline resources/fetchers/ecr_fetcher_test.go:215]: `s.Equal` should cover that

**Comment by ofiriro3** [comment]
> Still not sure I understand private/public separation as it seems like you are eventually merging the results. Definitely harder to understand when `ECRPublicProvider` is not implemented. Maybe we can find an easier way to merge the results, as it seems you have 2 entirely different implementations here. Let's discuss it again

You can have a short look over here -> 

https://github.com/aws/aws-sdk-go-v2/blob/main/service/ecr/api_op_DescribeRepositories.go

https://github.com/aws/aws-sdk-go-v2/blob/main/service/ecrpublic/api_op_DescribeRepositories.go

**Comment by ofiriro3** [comment]
What is the difference between Amazon ECR public and private repositories?
A private repository does not offer content search capabilities and requires Amazon IAM-based authentication using AWS account credentials before allowing images to be pulled. A public repository has descriptive content and allows anyone anywhere to pull images without needing an AWS account or using IAM credentials. Public repository images are also available in the Amazon ECR public gallery.

File: resources/fetchers/ecr_fetcher.go [hunk 2]
```diff
@@ -58,43 +62,42 @@ func (f *ECRFetcher) Stop() {
...
 	}
-
 	results = append(results, ECRResource{ecrRepositories})
-	return results, err
+
+	return results, nil
 }
 
-func (f *ECRFetcher) getAwsPodRepositories(ctx context.Context) ([]string, error) {
+func (f *ECRFetcher) describeAwsRepositories(ctx context.Context, describer PodDescriber) ([]ecr.Repository, error) {
```
[the remark this answers is not in the document]
Comment (ofiriro3) [reply resources/fetchers/ecr_fetcher.go:79]: Isn't describeAwsRepositories fit here?

File: resources/fetchers/ecr_fetcher.go [hunk 3]
```diff
@@ -58,43 +63,44 @@ func (f *ECRFetcher) Stop() {
...
+	return results, nil
 }
 
-func (f *ECRFetcher) getAwsPodRepositories(ctx context.Context) ([]string, error) {
-	podsList, err := f.kubeClient.CoreV1().Pods("").List(ctx, metav1.ListOptions{})
-	if err != nil {
-		logp.Error(fmt.Errorf("failed to get pods  - %w", err))
-		return nil, err
-	}
+func (f *ECRFetcher) describeAwsRepositories(ctx context.Context, podsList *v1.PodList, describer PodDescriber) ([]ecr.Repository, error) {
```
Comment (amirbenun) [inline resources/fetchers/ecr_fetcher.go:86]: Not a must only if you agree, I would rename it to `describePodImagesRepositories`

File: resources/fetchers/ecr_fetcher.go [hunk 4]
```diff
@@ -20,26 +20,31 @@ package fetchers
...
-	kubeClient        k8s.Interface
-	repoRegexMatchers []*regexp.Regexp
+	log           *logp.Logger
+	cfg           ECRFetcherConfig
+	kubeClient    k8s.Interface
+	PodDescribers []PodDescriber
+}
+
+type PodDescriber struct {
+	DescriberRegex *regexp.Regexp
```
Comment (amirbenun) [inline resources/fetchers/ecr_fetcher.go:46]: Again, not a must `FilterRegex`

File: resources/fetchers/ecr_fetcher.go [hunk 5]
```diff
@@ -20,26 +20,31 @@ package fetchers
...
-	ecrProvider       awslib.EcrRepositoryDescriber
-	kubeClient        k8s.Interface
-	repoRegexMatchers []*regexp.Regexp
+	log           *logp.Logger
+	cfg           ECRFetcherConfig
+	kubeClient    k8s.Interface
+	PodDescribers []PodDescriber
+}
+
+type PodDescriber struct {
```
Comment (amirbenun) [inline resources/fetchers/ecr_fetcher.go:45]: Not a must, `PartialPodDescriber`

**Review by amirbenun — APPROVED** [review]

File: resources/fetchers/ecr_fetcher.go (hunk 5 above)
Comment (ofiriro3) [reply resources/fetchers/ecr_fetcher.go:45]: Not sure what you meant by `partial` :/

I kept the name as it was for now

**Pull request merged**